### PR TITLE
Change DesiredNode version type to String

### PR DESCRIPTION
**Pull request by ldematte**
Changes the type of version in `DesiredNode` from `Version` to `String`. 

Comparison is by equality only, which modifies slightly the behaviour of `DesiredNodesSettingsValidator`, making it potentially stricter (e.g. rejecting requests from clients that specify a version newer (different) from the cluster).

### CODE REVIEW

**Comment by ldematte** [comment]
There are 3 things that makes this PR more complex:
1. docs. Code snippetes in docs are run as YAML rest tests; docs can specify a `{version}` substitution (see `DocsTestPlugin.groovy`) which does a round-trip string-version-string to get rid of any qualifiers. I updated the docs to use `{version_qualified}`, which takes the version as-is and does not try to parse it. We might want to check other usages and/or involve es-delivery in updating `DocsTestPlugin.groovy`
2. rest tests and current API. Yaml rest tests use some do and set sections to read data from the test cluster and initialize variables with it. Both `nodes.info` and `/_cat/nodes` still return `Version`. I had to relax the constraint checks to allow for differences.
3. BWC. Again, mismatch between the client and the test cluster; this time the test cluster is "old" (e.g. 8.10.3) and seems is queried with a 8.11.0-SNAPSHOT test. This makes the BWC tests fail. Since we want an equality comparison, this is probably expected/needed. What should be done to remove these tests from the BWC ones?

**Comment by ldematte** [comment]
@rjernst , regarding point 3 above, is it OK to remove these tests from BWC? (but leave them in CI - how do we do it?)

But even more, give there is just 1 equality check in the validator, I was wondering why we have it: do we need `node_version` in DesiredNode at all? Would it be better to replace it with an internal version (a "revision", progressive number), or with nothing at all and rely on RestApiVersion if/when we'll need it?

File: server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java [hunk 1]
```diff
@@ -23,19 +23,26 @@
 public class DesiredNodesSettingsValidator {
     private record DesiredNodeValidationError(int position, @Nullable String externalId, RuntimeException exception) {}
 
+    @Deprecated(forRemoval = true)
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java:26]: What is this going to be replaced with?

File: server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java [hunk 2]
```diff
@@ -23,19 +23,26 @@
...
+    private static boolean isCurrentIgnoringQualifiers(String expected, String actual) {
+        // The expected version string may contain -SNAPSHOT; the actual may not
+        // If the actual contains it, the expected must match entirely; this means actual must be a prefix for expected
+        return Strings.isNullOrEmpty(actual) == false && expected.startsWith(actual);
```
Comment (thecoop) [inline server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java:30]: Could use `hasText` here so we don't have the negation

File: server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java (hunk 1 above)
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java:26]: Hopefully, strict equality (`String.equals`)

File: server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java (hunk 2 above)
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java:30]: TIL, thanks

File: server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java (hunk 1 above)
Comment (ldematte) [reply server/src/main/java/org/elasticsearch/cluster/desirednodes/DesiredNodesSettingsValidator.java:26]: Or, based on the answers to [this question above](https://github.com/elastic/elasticsearch/pull/100133#issuecomment-1743392719), with nothing at all (no checks)

**Comment by ldematte** [comment]
After a long conversation with es-distrib (https://elastic.slack.com/archives/CC89MBSTA/p1696493952816229 - thank you  @fcofdez for taking all the time!) we concluded:
- For short term, we can deprecate the `version` field and remove the check. The check will be performed by ECK instead (by querying the cluster version and ensuring >=)
- We need to figure out a good way to express this "not downgrade" concept, which does not require semantic version comparison - this API will eventually be needed for ESS too.
- We will use knowledge we are accumulating in related work (e.g. Mapping allocation decisions) to find a better solution.

For now this PR is not needed anymore; we will need to open a new one to remove the check when `version` is deprecated.

**Pull request closed without merging**